util/cuda_timer_t.cuh

```diff
@@ -0,0 +1,75 @@
+// Copyright Supranational LLC
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef __SPPARK_UTIL_CUDA_TIMER_T_CUH__
+#define __SPPARK_UTIL_CUDA_TIMER_T_CUH__
+
+#include <iostream>
+#include <iomanip>
+
+class cuda_timer_t {
+    cudaEvent_t begin, end;
+public:
+    cuda_timer_t() : begin(nullptr), end(nullptr) {
+        CUDA_OK(cudaEventCreate(&begin));
+        CUDA_OK(cudaEventCreate(&end));
+    }
+
+    ~cuda_timer_t() {
+        if (begin) cudaEventDestroy(begin);
+        if (end) cudaEventDestroy(end);
+    }
+
+    inline void start() {
+        CUDA_OK(cudaEventRecord(begin, 0));
+    }
+
+    inline void start(cudaStream_t stream) {
+        CUDA_OK(cudaEventRecord(begin, stream));
+    }
+
+    inline void stop(const std::string str) {
+        float elapsed;
+
+        CUDA_OK(cudaEventRecord(end, 0));
+        CUDA_OK(cudaEventSynchronize(end));
+        CUDA_OK(cudaEventElapsedTime(&elapsed, begin, end));
+
+        std::cout << str << ": " << std::fixed << std::setprecision(3)
+                  << elapsed << " ms" << std::endl;
```

Comment:
This is rude to the application. Changing the stream formatting as suggested that is. The way to think about it is that you are providing interface for somebody else, not for yourself. If **your** method requires specific formatting, then save the stream state and restore it after you're done.

On a more general note I'd challenge the versatility of this method. I would find ability to "print" the timer more natural, as in `std::cout << my_timer << std::endl`. This naturally would require saving the stream value in the structure, so that `operator<<` would "know" which stream to record. In most general case it's formally unsafe, because you don't actually know if the stream would be destroyed by the caller, but it's reasonable to assume that it's not, especially in gpu_t.cuh context.

Reply:
On a second thought one can also question the versatility of std::fixed in combination with the suggested precision. While sub-millisecond kernels are likely a mistake, one can still argue that it would be appropriate to provide more usable printout for those. Say if the value is less than 0.05, multiply by 1000 and print "μs" instead of "ms"?

Reply:
> While sub-millisecond kernels are likely a mistake, one can still argue that it would be appropriate to provide more usable printout for those.

Following this suggestion, I also added a more usable printout for values `>= 1000` ms. My reasoning being that one might use the timer to time a group of operations where the total time of those might add up to more than a second.